pkg/reduce/data_forward.go

```diff
@@ -81,164 +97,515 @@ func NewDataForward(ctx context.Context,
+		windower:              windowingStrategy,
+		keyed:                 vertexInstance.Vertex.Spec.UDF.GroupBy.Keyed,
+		idleManager:           idleManager,
+		pbqManager:            pbqManager,
+		whereToDecider:        whereToDecider,
+		udfInvocationTracking: make(map[partition.ID]*pnf.ForwardTask),
+		of:                    of,
+		wmbChecker:            wmb.NewWMBChecker(2), // TODO: make configurable
+		log:                   logging.FromContext(ctx),
+		opts:                  options}
+
+	return rl, nil
 }
 
-// Start starts forwarding messages to readloop
-func (d *DataForward) Start() {
+// Start starts reading messages from ISG
+func (df *DataForward) Start() {
 	for {
 		select {
-		case <-d.ctx.Done():
-			d.log.Infow("Stopping reduce data forwarder...")
-
-			if err := d.fromBuffer.Close(); err != nil {
-				d.log.Errorw("Failed to close buffer reader, shutdown anyways...", zap.Error(err))
-			} else {
-				d.log.Infow("Closed buffer reader", zap.String("bufferFrom", d.fromBuffer.GetName()))
-			}
+		case <-df.ctx.Done():
 
 			// hard shutdown after timeout
 			cctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 			defer cancel()
-			// allow readloop to clean itself up.
-			d.readloop.ShutDown(cctx)
-
-			for _, v := range d.toBuffers {
-				if err := v.Close(); err != nil {
-					d.log.Errorw("Failed to close buffer writer, shutdown anyways...", zap.Error(err), zap.String("bufferTo", v.GetName()))
-				} else {
-					d.log.Infow("Closed buffer writer", zap.String("bufferTo", v.GetName()))
-				}
-			}
+			// allow to clean itself up.
+			df.ShutDown(cctx)
 
-			// stop watermark fetcher
-			if err := d.watermarkFetcher.Close(); err != nil {
-				d.log.Errorw("Failed to close watermark fetcher", zap.Error(err))
-			}
-
-			// stop watermark publisher
-			for _, publisher := range d.watermarkPublishers {
-				if err := publisher.Close(); err != nil {
-					d.log.Errorw("Failed to close watermark publisher", zap.Error(err))
-				}
-			}
 			return
 		default:
-			// pass the child context so that the reader can be closed before the readloop
+			// pass the child context so that the reader can be closed.
 			// this way we can avoid the race condition and have all the read messages persisted
 			// and acked.
-			d.forwardAChunk(d.ctx)
+			df.forwardAChunk(df.ctx)
 		}
 	}
 }
 
+// ReplayPersistedMessages replays persisted messages, because during boot up, it has to replay the data from the persistent store of
+// PBQ before it can start reading from ISB. ReplayPersistedMessages will return only after the replay has been completed.
+func (df *DataForward) ReplayPersistedMessages(ctx context.Context) error {
+	// start the PBQManager which discovers and builds the state from persistent store of the PBQ.
+	partitions, err := df.pbqManager.GetExistingPartitions(ctx)
+	if err != nil {
+		return err
+	}
+
+	df.log.Infow("Partitions to be replayed ", zap.Int("count", len(partitions)), zap.Any("partitions", partitions))
```

Comment:
better to put `zap.Any("partitions", partitions)` into separate Debug level. sliding window, etc could have 10s of thousands here.